File: app/src/main/java/com/alex/miet/mobile/data/schedule/ScheduleRepositoryImpl.java
package com.alex.miet.mobile.data.schedule;

import java.util.List;

import com.alex.miet.mobile.ScheduleModel;
import com.alex.miet.mobile.ScheduleModelDao;

/**
 * Specific {@link ScheduleRepository} implementation
 */
public class ScheduleRepositoryImpl implements ScheduleRepository {

    private final ScheduleModelDao dao;

    public ScheduleRepositoryImpl(ScheduleModelDao dao) {
        this.dao = dao;
    }

    @Override
    public void save(ScheduleModel entity) {
        dao.save(entity);
    }

    @Override
    public void saveAll(Iterable<ScheduleModel> entities) {
        dao.saveInTx(entities);
    }

    @Override
    public List<ScheduleModel> getAll() {
        return dao.loadAll();
    }

    @Override
    public void update(ScheduleModel entity) {
        dao.update(entity);
    }

    @Override
    public void updateAll(Iterable<ScheduleModel> entities) {
        dao.updateInTx(entities);
    }

    @Override
    public void delete(ScheduleModel entity) {
        dao.delete(entity);
    }

    @Override
    public void deleteAll() {
        dao.deleteAll();
    }

    @Override
    public ScheduleModel getGroupByName(String groupName) {
        return dao.queryBuilder().where(ScheduleModelDao.Properties.Group.eq(groupName)).build()
                .unique();
    }

    @Override
    public void replaceByGroupName(String groupName, ScheduleModel model) {
        deleteByGroupName(groupName);
        dao.insertOrReplace(model);
    }

    @Override
    public void deleteByGroupName(String groupName) {
        ScheduleModel scheduleByGroup = dao.queryBuilder()
                .where(ScheduleModelDao.Properties.Group.eq(groupName)).build().unique();
        if (scheduleByGroup != null) {
            delete(scheduleByGroup);
        }
    }

    @Override
    public boolean isGroupCached(String groupName) {
        return getGroupByName(groupName) != null;
    }
}

File: app/src/main/java/com/alex/miet/mobile/data/schedule/ScheduleRepositoryImpl.kt
package com.alex.miet.mobile.data.schedule

import com.alex.miet.mobile.daos.GroupDao
import com.alex.miet.mobile.entities.GroupItem
import io.reactivex.Maybe

/**
 * Specific [ScheduleRepository] implementation
 */
class ScheduleRepositoryImpl(private val dao: GroupDao) : ScheduleRepository {

    override fun save(entity: GroupItem) {
        dao.insert(entity)
    }

    override fun saveAll(entities: List<GroupItem>) {
        dao.insert(entities)
    }

    override fun getAll(): Maybe<List<GroupItem>> {
        return dao.loadAll()
    }

    override fun update(entity: GroupItem) {
        dao.update(entity)
    }

    override fun updateAll(entities: List<GroupItem>) {
        dao.update(entities)
    }

    override fun delete(entity: GroupItem) {
        dao.delete(entity)
    }

    override fun deleteAll() {
        dao.deleteAll()
    }

    override fun getGroupByName(groupName: String): Maybe<GroupItem> {
        return dao.getGroupByName(groupName)
    }

    override fun replaceByGroupName(groupName: String, model: GroupItem) {
        deleteByGroupName(groupName)
        dao.insert(model)
    }

    override fun deleteByGroupName(groupName: String) {
        val scheduleByGroup = dao.getGroupByName(groupName).blockingGet()
        if (scheduleByGroup != null) {
            delete(scheduleByGroup)
        }
    }

    override fun isGroupCached(groupName: String): Boolean {
        return dao.getGroupByName(groupName).blockingGet()?.group == groupName
    }
}
